flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogBase.java

```diff
@@ -407,4 +443,201 @@ private Table getHiveTable(ObjectPath tablePath) throws TableNotExistException {
+		}
+	}
+
+	@Override
+	public void alterPartition(ObjectPath tablePath, CatalogPartitionSpec partitionSpec, CatalogPartition newPartition, boolean ignoreIfNotExists)
+		throws PartitionNotExistException, CatalogException {
+		// Explicitly check if the partition exists or not
+		// because alter_partition() doesn't throw NoSuchObjectException like dropPartition() when the target doesn't exist
+		if (partitionExists(tablePath, partitionSpec)) {
+			try {
+				Table hiveTable = getHiveTable(tablePath);
+				ensurePartitionedTable(tablePath, hiveTable);
+				client.alter_partition(
+					tablePath.getDatabaseName(),
+					tablePath.getObjectName(),
+					createHivePartition(hiveTable, partitionSpec, newPartition)
+				);
+			} catch (TException e) {
+				throw new CatalogException(
+					String.format("Failed to alter existing partition with new partition %s of table %s",
+						partitionSpec, tablePath), e);
+			} catch (TableNotPartitionedException | TableNotExistException | PartitionSpecInvalidException e) {
+				throw new PartitionNotExistException(catalogName, tablePath, partitionSpec, e);
+			}
+		} else if (!ignoreIfNotExists) {
+			throw new PartitionNotExistException(catalogName, tablePath, partitionSpec);
+		}
+	}
+
+	private void ensurePartitionedTable(ObjectPath tablePath, Table hiveTable) throws TableNotPartitionedException {
+		if (hiveTable.getPartitionKeysSize() == 0) {
+			throw new TableNotPartitionedException(catalogName, tablePath);
+		}
+	}
+
+	/**
+	 * Get field names from field schemas.
+	 */
+	static List<String> getFieldNames(List<FieldSchema> fieldSchemas) {
+		List<String> names = new ArrayList<>(fieldSchemas.size());
+		for (FieldSchema fs : fieldSchemas) {
+			names.add(fs.getName());
+		}
+		return names;
+	}
+
+	/**
+	 * Creates a {@link CatalogPartitionSpec} from a Hive partition name string.
+	 * Example of Hive partition name string - "name=bob/year=2019"
+	 */
+	private static CatalogPartitionSpec createPartitionSpec(String hivePartitionName) {
+		String[] partKeyVals = hivePartitionName.split("/");
+		Map<String, String> spec = new HashMap<>(partKeyVals.length);
+		for (String keyVal : partKeyVals) {
+			String[] kv = keyVal.split("=");
+			spec.put(kv[0], kv[1]);
+		}
+		return new CatalogPartitionSpec(spec);
+	}
+
+	/**
+	 * Get a list of ordered partition values by re-arranging them based on the given list of partition keys.
+	 *
+	 * @param partitionSpec a partition spec.
+	 * @param partitionKeys a list of partition keys.
+	 * @return A list of partition values ordered according to partitionKeys.
+	 * @throws PartitionSpecInvalidException thrown if partitionSpec and partitionKeys have different sizes,
+	 *                                       or any key in partitionKeys doesn't exist in partitionSpec.
+	 */
+	List<String> getFullPartitionValues(ObjectPath tablePath, CatalogPartitionSpec partitionSpec, List<String> partitionKeys)
```

Comment:
what does the "full" mean here? probably rename to "getOrdered/ArrangedPartitionValues" to conform to its logic?

nit: can we move `tablePath` to the end as the last param given it's here to only build a message without any real effect.

Reply:
It means a full spec, which contains values for all partition keys. And a partial spec can only contain values for a subset of partition keys. Operations like `createPartition`, `dropPartition` require a full spec. Operations like `listPartitions` can accept partial spec. Let me rename to to getOrderedFullPartitionValues, to indicate it orders the values and requires a full spec.